### PR TITLE
SSRF Routes

### PR DESCRIPTION
**Pull request by meandthemoon**
Sink code and the new page's content are in new NPM packages:
[test-bench-utils](https://www.npmjs.com/package/@contrast/test-bench-utils)
[test-bench-content](https://www.npmjs.com/package/@contrast/test-bench-content)

Vulnerable routes currently support GET and POST requests.

### CODE REVIEW

**Review by Avuxo — APPROVED** [review]

**Review by cixel — APPROVED** [review]

File: routes/ssrf.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    ctx.render('../node_modules/@contrast/test-bench-content/views/ssrf', {
+      requestUrl: EXAMPLE_URL
+    })
+  );
+
+  const libs = ['axios', 'bent', 'fetch', 'request', 'superagent'];
+
+  libs.forEach((lib) => {
+    router.get(`/ssrf/${lib}/unsafe`, async (ctx, next) => {
+      const url = createUnsafeUrl(ctx.query.input);
```
Comment (cixel) [inline routes/ssrf.js:20]: ~This needs to be `query['untrusted-input']` based on the key used in the UI.~

File: routes/ssrf.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      const data = await makeRequest(lib, url);
+
+      ctx.body = data;
+    });
+
+    router.post(`/ssrf/${lib}/unsafe`, async (ctx, next) => {
+      const url = createUnsafeUrl(ctx.request.body.input);
```
Comment (cixel) [inline routes/ssrf.js:27]: ~This needs to be `body['untrusted-input']` based on the key used in the UI.~

File: routes/ssrf.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    router.get(`/ssrf/${lib}/safe`, async (ctx, next) => {
+      const url = createSafeUrl(ctx.query.input);
+      const data = await makeRequest(lib, url);
+
+      ctx.body = data;
+    });
+
+    router.post(`/ssrf/${lib}/safe`, async (ctx, next) => {
+      const url = createSafeUrl(ctx.request.body.input);
```
Comment (cixel) [inline routes/ssrf.js:41]: ~This needs to be `body['untrusted-input']` based on the key used in the UI.~

File: routes/ssrf.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    router.post(`/ssrf/${lib}/unsafe`, async (ctx, next) => {
+      const url = createUnsafeUrl(ctx.request.body.input);
+      const data = await makeRequest(lib, url);
+
+      ctx.body = data;
+    });
+
+    router.get(`/ssrf/${lib}/safe`, async (ctx, next) => {
+      const url = createSafeUrl(ctx.query.input);
```
Comment (cixel) [inline routes/ssrf.js:34]: This needs to be `query['untrusted-input']` based on the key used in the UI.

**Review by cixel — CHANGES_REQUESTED** [review]

File: package.json [hunk 5]
```diff
@@ -12,6 +12,10 @@
   "author": "Contrast Security",
   "license": "MIT",
   "dependencies": {
+    "@contrast/test-bench-content": "^1.0.0",
+    "@contrast/test-bench-utils": "^1.0.0",
+    "axios": "^0.19.0",
```
Comment (bizob2828) [inline package.json]: the request libs are a dep of the test-bench-utils, we don't need them here do we?

File: routes/ssrf.js (hunk 4 above)
Comment (meandthemoon) [reply routes/ssrf.js:34]: Isn't 'input' more idiomatic for our use?

File: routes/ssrf.js (hunk 3 above)
Comment (meandthemoon) [reply routes/ssrf.js:41]: Isn't 'input' more idiomatic for our use?

File: routes/ssrf.js (hunk 4 above)
Comment (cixel) [reply routes/ssrf.js:34]: Yeah, but it's not what the `@contrast/test-bench-content` page is using. Either we should update that to send `input` or update what all the APIs do.

Comment (meandthemoon) [reply routes/ssrf.js:34]: I'd prefer to update the template unless you have serious concerns?

**Review by cixel — APPROVED** [review]

File: routes/ssrf.js (hunk 4 above)
Comment (meandthemoon) [reply routes/ssrf.js:34]: I've updated the test-bench-content repo to use `name='input'` in the markup.

**Review by bizob2828 — CHANGES_REQUESTED** [review]

File: routes/ssrf.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+const {
+  sinks: { ssrf }
+} = require('@contrast/test-bench-utils');
+
+const EXAMPLE_URL = 'http://www.example.com';
+
+module.exports = ({ router }) => {
+  router.get('/ssrf', (ctx) =>
+    ctx.render('../node_modules/@contrast/test-bench-content/views/ssrf', {
```
Comment (bizob2828) [inline routes/ssrf.js]: this is unfortunate. there has to be a way to get that string from requiring of the test-bench-content module

Comment (bizob2828) [reply routes/ssrf.js]: looks like maybe we can use `renderFile`?

Comment (bizob2828) [reply routes/ssrf.js]: oh that's not even an npm module

File: package.json (hunk 5 above)
Comment (meandthemoon) [reply package.json]: Aye, thanks

File: routes/ssrf.js (hunk 6 above)
Comment (meandthemoon) [reply routes/ssrf.js]: This was meant as a temporary solution until _all_ shared templates lived in the new repo and we could then configure the view engine to use the package folder as the source of views. I agree that this isn't great. I can continue and see if there's a better short-term solution.

File: package.json (hunk 5 above)
Comment (meandthemoon) [reply package.json]: The request libs are now removed.

File: routes/ssrf.js (hunk 6 above)
Comment (meandthemoon) [reply routes/ssrf.js]: This was resolved. A rule-specific template was installed that pulled in the view from the template. This way is at least an improvement I think.

**Review by bizob2828 — APPROVED** [review]

**Comment by bizob2828** [comment]
You do have conflicts but besides that LGTM

**Pull request merged**